src/main/java/io/github/zap/commons/event/BukkitProxy.java

```diff
@@ -0,0 +1,116 @@
+package io.github.zap.commons.event;
+
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.Listener;
+import org.bukkit.plugin.EventExecutor;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.RegisteredListener;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Objects;
+import java.util.logging.Level;
+
+class BukkitProxy<T extends org.bukkit.event.Event> extends SimpleEvent<T> {
+    private final Listener listener = new Listener() {};
+
+    private final Plugin plugin;
+    private final Class<T> bukkitEventClass;
+    private final EventPriority priority;
+    private final boolean ignoreCancelled;
+
+    private boolean eventRegistered = false;
+    private boolean reflectionFailed = false;
+    private HandlerList handlerList;
+    private RegisteredListener registeredListener;
+
+    BukkitProxy(@NotNull Plugin plugin, @NotNull Class<T> bukkitEventClass, @NotNull EventPriority priority,
+                boolean ignoreCancelled) {
+        this.plugin = Objects.requireNonNull(plugin, "plugin cannot be null");
+        this.bukkitEventClass = Objects.requireNonNull(bukkitEventClass, "bukkitEventClass cannot be null");
+        this.priority = Objects.requireNonNull(priority, "priority cannot be null");
+        this.ignoreCancelled = ignoreCancelled;
+    }
+
+    private void maybeReflectHandlerList() {
```

Comment:
I think this should return `HanderList` instead of void for a bit better thread safety (caller don't have to rely on this method side effect)

Reply:
thread safety should be accomplished by calling synchronized() on the bukkitproxy, which covers every method (including these)

Reply:
well I just take thread-safety as an example, but it is still better for the method to return a value that can be used like `HandlerList hl = maybeReflectHandlerList();` rather than using the class field

Reply:
i think it makes no practical difference in this situation but i have updated

Reply:
I just prefer getters :)